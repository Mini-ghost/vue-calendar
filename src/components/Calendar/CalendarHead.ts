import Vue from 'vue'

// Types
import type { VNode } from 'vue'
import type { PropValidator } from 'vue/types/options'

import { fill } from '@/utils/fill'
import { genFormatter } from '@/utils/genFormatter'

const formatter = genFormatter('en-US', {
  year: 'numeric',
  month: 'long'
})

export default Vue.extend({
  name: 'CalendarHead',
  props: {
    value: {
      type: String,
      required: true
    } as PropValidator<string>
  },

  computed: {
    valueYear() {
      return Number(this.value.split('-')[0])
    },
    valueMonth() {
      return Number(this.value.split('-')[1]) - 1
    },
  },

  methods: {
    genHeader(): VNode {
      return this.$createElement('span', [formatter(this.value)])
    },
    genBtn(change: number): VNode {
      return this.$createElement('button', {
        on: {
          click: (e: Event) => {
            e.stopPropagation()

            let year: number
            let month: number
            
            if(change > 0) {
              year = this.valueMonth === 11
                ? this.valueYear + 1 
                : this.valueYear

              month = (this.valueMonth + 1) % 12
            } else {
              year = this.valueMonth ? this.valueYear : this.valueYear - 1
              month = (this.valueMonth + 11) % 12
            }

            const date = `${year}-${fill(month + 1)}`
            this.$emit('input', date)
          }
        }
      }, [change > 0 ? 'Next' : 'Prev'])
    }
  },  
  render (): VNode {
    return this.$createElement('div', {

    }, [
      this.genBtn(-1),
      this.genHeader(),
      this.genBtn(1)
    ])
  }
})