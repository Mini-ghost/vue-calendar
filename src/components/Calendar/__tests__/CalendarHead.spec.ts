import {
  mount,
  MountOptions,
  Wrapper,
} from '@vue/test-utils'

import CalendarHead from '../CalendarHead'

describe('CalendarHead.ts', () => {
  type Instance = InstanceType<typeof CalendarHead>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>
  
  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(CalendarHead, options)
    }
  })

  it('should trigger event on arrows click', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2021-01',
      },
    })

    const input = jest.fn()
    wrapper.vm.$on('input', input)

    wrapper.findAll('button').at(0).trigger('click')
    expect(input).toHaveBeenCalledWith('2020-12')

    wrapper.findAll('button').at(1).trigger('click')
    expect(input).toHaveBeenCalledWith('2021-02')
  })
})
