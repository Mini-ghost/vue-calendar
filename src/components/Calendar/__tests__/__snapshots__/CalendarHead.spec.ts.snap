// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHead.ts should render component and match snapshot 1`] = `
<div class="flex relative items-center py-2"><button class="flex justify-center items-center w-8 h-8 bg-transparent hover:bg-gray-200 rounded-full border-transparent hover:border-blue-100 cursor-pointer"><svg width="20px" height="20px" viewBox="0 0 20 20">
      <path d="M13.1,5.6c0,0.1,0,0.2-0.1,0.2L8.8,10l4.1,4.1c0.1,0.1,0.1,0.2,0.1,0.2c0,0.1,0,0.2-0.1,0.2l-0.5,0.5 c-0.1,0.1-0.2,0.1-0.2,0.1c-0.1,0-0.2,0-0.2-0.1L7,10.2C7,10.2,6.9,10.1,6.9,10S7,9.8,7,9.8l4.9-4.9c0.1-0.1,0.2-0.1,0.2-0.1 c0.1,0,0.2,0,0.2,0.1L13,5.4C13,5.5,13.1,5.5,13.1,5.6L13.1,5.6z"></path>
    </svg></button>
  <div class="flex-1 flex relative overflow-hidden">
    <transition-stub name="transition-slide">
      <div class="text-center w-full duration-300">May 2021</div>
    </transition-stub>
  </div><button class="flex justify-center items-center w-8 h-8 bg-transparent hover:bg-gray-200 rounded-full border-transparent hover:border-blue-100 cursor-pointer"><svg width="20px" height="20px" viewBox="0 0 20 20">
      <path d="M13.1,10c0,0.1,0,0.2-0.1,0.2l-4.9,4.9c-0.1,0.1-0.2,0.1-0.2,0.1c-0.1,0-0.2,0-0.2-0.1L7,14.6c-0.1-0.1-0.1-0.1-0.1-0.2 c0-0.1,0-0.2,0.1-0.2l4.1-4.1L7,5.9C7,5.8,6.9,5.7,6.9,5.6S7,5.5,7,5.4l0.5-0.5c0.1-0.1,0.2-0.1,0.2-0.1c0.1,0,0.2,0,0.2,0.1L13,9.8 C13,9.8,13.1,9.9,13.1,10L13.1,10z"></path>
    </svg></button>
</div>
`;
